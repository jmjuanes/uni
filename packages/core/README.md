# @unicss/core

A tiny and framework agnostic CSS-in-JS solution.

## Installation

Use **npm** or **yarn** for adding this package to your project:

```bash
$ yarn add @unicss/core
``` 

## Usage

```js
import {css} from "@unicss/core";

const button = css({
    backgroundColor: "#2271b1",
    borderRadius: "0.5rem",
    color: "white",
    padding: "0.5rem 1rem",
    "&:hover": {
        cursor: "pointer",
    },
});

document.getElementById("button").classList.add(button);
```

## API

### css(styles)

A function to generate a classname from the specified styles object.

```js
import {css} from "@unicss/core";

const name = css({
    color: "red",
    fontWeight: "bold",
    "&:hover": {
        color: "blue",
        cursor: "pointer",
    },
});
```

### globalCss(styles)

A function to convert the speicifed styles object to global styles. Useful for registering reset styles or font faces.

```js
import {globalCss} from "@unicss/core";

globalCss({
    body: {
        backgroundColor: "#fff",
        color: "#000",
    },
    a: {
        textDecoration: "none",
    },
});
```

### keyframes(obj)

Generate global animations from the specified keyframes configuration object. This function returns an unique animation identifier that can be used in an `animation` CSS property:

```js
import {keyframes, css} from "@unicss/core";

const scale = keyframes({
    from: {
        transform: "scale(1)",
    },
    to: {
        transform: "scale(1.5)",
    },
});

const image = css({
    height: "500px",
    width: "500px",
    "&:hover": {
        animation: `${scale} 1s ease`,
    },
});

document.getElementById("image").classList.add(image);
```

### createUni(config)

Use this function to generate a new instance of UniCSS if you need to provide your own theme.

```js
import {createUni} from "@unicss/core";

const {css, keyframes, globalCss} = createUni({
    theme: {
        colors: {
            primary: colors.mint["500"],
            secondary: colors.blue["500"],
        },
    },
    // ...other configuration
});
```

### classNames(obj)

A tiny utility for conditionally joining class names.

```js
import {classNames} from "@unicss/core";

const names = classNames({
    "foo": true,
    "bar": trueCondition === true,
    "baz": null,
});
// names === "foo bar"
```


# License

[MIT License](https://github.com/jmjuanes/unicss/blob/main/LICENSE).
