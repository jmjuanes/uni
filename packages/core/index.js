// Default theme scales mapping
export const defaultThemeMap = {
    backgroundColor: "colors",
    borderColor: "colors",
    borderBottomColor: "colors",
    borderLeftColor: "colors",
    borderRightColor: "colors",
    borderTopColor: "colors",
    borderRadius: "radius",
    borderBottomLeftRadius: "radius",
    borderBottomRightRadius: "radius",
    borderTopLeftRadius: "radius",
    borderTopRightRadius: "radius",
    borderWidth: "sizes",
    borderBottomWidth: "sizes",
    borderLeftWidth: "sizes",
    borderRightWidth: "sizes",
    borderTopWidth: "sizes",
    bottom: "spacing",
    boxShadow: "shadows",
    color: "colors",
    fill: "colors",
    fontFamily: "fonts",
    fontSize: "fontSizes",
    fontWeight: "fontWeights",
    height: "sizes",
    left: "spacing",
    lineHeight: "lineHeights",
    margin: "spacing",
    marginBottom: "spacing",
    marginLeft: "spacing",
    marginRight: "spacing",
    marginTop: "spacing",
    maxHeight: "sizes",
    maxWidth: "sizes",
    minHeight: "sizes",
    minWidth: "sizes",
    opacity: "opacities",
    padding: "spacing",
    paddingBottom: "spacing",
    paddingLeft: "spacing",
    paddingRight: "spacing",
    paddingTop: "spacing",
    right: "spacing",
    textShadow: "shadows",
    top: "spacing",
    width: "sizes",
};

const hashCode = str => {
    return "uni-" + Math.abs(Array.from(str).reduce((s, c) => Math.imul(31, s) + c.charCodeAt(0) | 0, 0)).toString();
};

// Exclude a field from the specified object
const exclude = (obj, field) => {
    return Object.fromEntries(Object.entries(obj).filter(e => e[0] !== field));
};

// Tiny reducer alias
const toObject = (list, fn) => list.reduce(fn, {});

// Get value in object from path
export const get = (obj, path) => {
    return path.split(".").reduce((o, p) => o?.[p], obj);
};

// Merge CSS styles
export const merge = (source, target) => {
    source = {...source};
    Object.keys(target).forEach(key => {
        // Check for @font-face attribute
        if (key === "@font-face") {
            if (typeof source["@font-face"] !== "undefined") {
                source[key] = [source[key]].flat(1).concat(target[key]).flat(1);
                return;
            }
        }
        // Check for object property --> deep merge
        else if (typeof source[key] === "object" && typeof target[key] === "object") {
            source[key] = merge(source[key], target[key]);
            return;
        }
        // Other value --> override source property
        source[key] = target[key];
    });
    return source;
};

// Wrap CSS Rule
const wrapRule = (ruleName, ruleContent, separator) => {
    return `${ruleName} {${separator || ""}${ruleContent}${separator || ""}}`;
};

// Parse CSS property
const parseProp = (prop, value, config) => {
    let propsToParse = {
        [prop]: value,
    };
    // Check if custom properties object has been provided
    if (typeof config?.properties?.[prop] === "function") {
        const newProps = config.properties[prop](value);
        if (newProps && typeof newProps === "object") {
            propsToParse = newProps;
        }
    }
    // Parse props
    return Object.keys(propsToParse).map(item => ([
        item.replace(/[A-Z]/g, letter => `-${letter.toLowerCase()}`),
        propsToParse[item],
    ]));
};

// Parse CSS value
const parseValue = (prop, value, config) => {
    if (typeof value === "string" && value.indexOf("$") > -1 && defaultThemeMap[prop]) {
        const scaleName = defaultThemeMap[prop];
        const scale = config?.theme?.[scaleName];

        if (scale && typeof scale === "object") {
            return value.replace(/\$([^\s]+)/g, (match, key) => {
                return typeof scale[key] !== "undefined" ? scale[key].toString() : match;
            });
        }
    }
    return value;
};

// Parse mixins
const parseMixins = (styles, config, prev) => {
    prev = prev || new Set();
    if (config?.mixins && styles?.apply && (typeof styles.apply === "string" || Array.isArray(styles.apply))) {
        const mixinsList = [styles.apply].flat().filter(n => {
            return n && typeof n === "string";
        });
        return {
            ...exclude(styles, "apply"),
            ...toObject(mixinsList, (newStyles, mixinName) => {
                if (prev.has(mixinName)) {
                    return newStyles;
                }
                // Apply styles from this mixin
                prev.add(mixinName);
                return {
                    ...newStyles,
                    ...parseMixins(get(config?.mixins || {}, mixinName), config, prev),
                };
            }),
        };
    }
    // No mixins to apply
    return styles || {};
};

// Transform a selector rule
const transformSelector = (selector, styles, config) => {
    if (styles && Array.isArray(styles)) {
        return styles.map(s => transformSelector(selector, s, config)).flat();
    }
    const result = [""];
    Object.entries(parseMixins(styles, config)).forEach(([key, value]) => {
        // skip content in the 'variants' key: reserved only to register variants
        // Skip content in the 'apply' key: reserved only to mixins
        if (key === "variants" || key === "apply" || value === null) {
            return;
        }
        // Nested styles or @media rules 
        if (typeof value === "object") {
            // Media rules styles should be wrapped into a new rule
            if (/^@/.test(key)) {
                return result.push(
                    wrapRule(key, transformSelector(selector, value, config), "\n"),
                );
            }
            // Add nested styles
            // We should replace the & with the current selector
            return result.push(
                transformSelector(key.replaceAll("&", selector), value, config),
            );
        }
        // Just parse as a simple property
        parseProp(key, value, config).forEach(([prop, value]) => {
            result[0] = result[0] + `${prop}:${parseValue(prop, value, config)};`;
        });
    });
    // result[0] contains the styles for the current selector, and should be wrapped into {}
    result[0] = wrapRule(selector, result[0], "");
    return result.flat(2);
};

// Transform a styles object to string
export const transform = (styles, config) => {
    const result = Object.entries(styles || {}).map(([key, value]) => {
        // Check for at rule
        if (key.startsWith("@")) {
            // Check for @import rule, just add one @import for each source
            // value must be a string or an array of strings
            if (key === "@import") {
                return [value].flat().map(v => `@import ${v};`);
            }
            // Check for @font-face rule
            else if (["@fontFace", "@fontface", "@font-face"].indexOf(key) > -1) {
                return transformSelector("@font-face", value, config);
            }
            // if (/^@(media|keyframes)/.test(key.trim())) {
            else {
                // Other rule (for example @media or @keyframes)
                const rules = Object.keys(value).map(k => {
                    return transformSelector(k, value[k], config);
                });
                return wrapRule(key, rules.join(" "));
            }
        }
        // Other value --> parse as regular classname
        return transformSelector(key, value, config);
    });
    return result.flat().join("\n");
};

const registerStyles = (hash, css, cache, target) => {
    if (!cache.has(hash)) {
        target.innerHTML = target.innerHTML + css.replaceAll("__uni__", hash) + "\n";
        cache.add(hash);
    }
    return hash;
};

const createTarget = key => {
    // Check if we are in a DOM env
    if (typeof document !== "undefined") {
        let target = document.querySelector(`[data-source="uni/${key}"`);
        if (!target) {
            target = document.createElement("style");
            target.dataset.source = `uni/${key}`;
            target.innerHTML = "";
            document.head.appendChild(target);
        }
        return target;
    }
    // Other case: we are not in an env supporting DOM (maybe SSR)
    // return a fake <style> tag
    return {
        innerHTML: "",
    };
};

// Create a new instance of UniCSS
export const createUni = config => {
    const cache = new Set();
    config = config || {};
    config.target = config.target || createTarget("css");

    // Register css styles
    const css = s => {
        const styles = transformSelector(".__uni__", s, config).join("\n"); 
        const hash = hashCode(styles);
        return registerStyles(hash, styles, cache, config.target);
    };

    // Generate a keyframes styles
    const keyframes = s => {
        const styles = wrapRule(
            "@keyframes __uni__",
            Object.keys(s).map(k => transformSelector(k, s[k], config)).join(" "),
        );
        const hash = hashCode(styles);
        return registerStyles(hash, styles, cache, config.target);
    };

    // Generate global styles
    const globalCss = s => {
        const styles = transform(s, config);
        const hash = hashCode(styles);
        return registerStyles(hash, styles, cache, config.target);
    };

    return {
        css,
        globalCss,
        keyframes,
        extractCss: () => config.target?.innerHTML || "",
        theme: config?.theme || {},
        target: config?.target,
    };
};

// Cached instance
let cachedInstance = null;
const getCachedInstance = () => {
    return cachedInstance || (cachedInstance = createUni({}));
};

export const css = s => getCachedInstance().css(s);
export const globalCss = s => getCachedInstance().globalCss(s);
export const keyframes = s => getCachedInstance().keyframes(s);
export const extractCss = () => getCachedInstance().extractCss();

// Tiny utility for conditionally joining classNames
export const classNames = (...args) => {
    return (args || []).map(items => {
        if (typeof items === "string") {
            return items.split(" ").filter(item => item.length);
        }
        else if (Array.isArray(items)) {
            return items.filter(item => typeof item === "string" && item.length); 
        }
        else if (typeof items === "object") {
            return Object.keys(items || {}).filter(key => !!items[key]);
        }
        return [];
    }).flat().join(" ");
};
