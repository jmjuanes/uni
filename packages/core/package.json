{
    "name": "@unicss/core",
    "version": "0.0.2",
    "description": "The framework-agnostic CSS-in-JS implementation",
    "author": "Josemi Juanes <hello@josemi.xyz>",
    "type": "module",
    "license": "MIT",
    "repository": "https://github.com/jmjuanes/unicss",
    "bugs": "https://github.com/jmjuanes/unicss/issues",
    "homepage": "https://www.josemi.xyz/unicss",
    "exports": "./index.js",
    "keywords": [
        "unicss",
        "core",
        "css-in-js"
    ]
}