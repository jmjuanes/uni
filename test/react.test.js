import renderer, { create } from "react-test-renderer";
import {createCache} from "@unicss/core";
import {
    CacheProvider,
    useCache,
    ThemeProvider,
    useTheme,
    styled,
    Box,
} from "@unicss/react";
 
const selector = `style[data-source="uni/react"]`;

describe("[react] ThemeProvider", () => {
    it("should render", () => {
        const component = renderer.create((
            <ThemeProvider theme={{}}>
                <div>Hello world</div>
            </ThemeProvider>
        ));

        expect(component.toJSON()).toMatchSnapshot();
    });
});

describe("[react] Box", () => {
    it("should render", () => {
        const component = renderer.create((
            <ThemeProvider theme={{}}>
                <Box css={{color: "white"}}>Hello world</Box>
            </ThemeProvider>
        ));

        expect(component.toJSON()).toMatchSnapshot();
    });
});

describe("[react] useTheme", () => {
    it("should return current theme", () => {
        const defaultTheme = {
            colors: {
                primary: "black",
            },
        };
        let theme = null;
        let TestComponent = () => {
            theme = useTheme();
            return null;
        };
        renderer.create((
            <ThemeProvider theme={defaultTheme}>
                <TestComponent />
            </ThemeProvider>
        ));

        expect(theme).not.toBeNull();
        expect(theme?.colors?.primary).toBe(defaultTheme.colors.primary);
    });
});

describe("[react] styled", () => {
    it("should return a valid React component", () => {
        const StyledComponent = styled("div", {
            color: "white",
        });
        const component = renderer.create((
            <ThemeProvider theme={{}}>
                <StyledComponent>Hello world</StyledComponent>
            </ThemeProvider>
        ));

        expect(component.toJSON()).toMatchSnapshot();
    });

    it("should apply theme", () => {
        const theme = {
            scales: {
                colors: {
                    primary: "black",
                    secondary: "blue",
                },
            },
        };
        const StyledComponent = styled("div", {
            backgroundColor: "secondary",
            color: "primary",
        });
        const component = renderer.create((
            <ThemeProvider theme={theme}>
                <StyledComponent>Hello</StyledComponent>
            </ThemeProvider>
        ));
        const div = component.root.findByType("div");

        expect(div).toBeDefined();
        expect(div.props.className).toEqual(expect.stringContaining("uni-"));
        expect(document.querySelector(selector).innerHTML).toEqual(expect.stringContaining("color:black"));
    });
});

describe("[react] CacheProvider", () => {
    it("should render", () => {
        const component = renderer.create((
            <CacheProvider>
                <div>Hello world</div>
            </CacheProvider>
        ));

        expect(component.toJSON()).toMatchSnapshot();
    });

    it("should provide cache to ThemeProvider", () => {
        const cache = createCache({key: "react-test"});
        renderer.create((
            <CacheProvider cache={cache}>
                <ThemeProvider theme={{}}>
                    <Box css={{backgroundColor: "mint"}} />
                </ThemeProvider>
            </CacheProvider>
        ));

        expect(document.querySelector("style[data-source='uni/react-test']")).toBeDefined();
        // expect(cache.target.innerHTML).not.toBe("");
        expect(cache.target.innerHTML).toEqual(expect.stringContaining(".uni-"));
    });
});

describe("[react] useCache", () => {
    it("should return provided cache", () => {
        const cache = createCache();
        let providedCache = null;
        const InternalComponent = () => {
            providedCache = useCache();
            return null;
        };
        const component = renderer.create((
            <CacheProvider cache={cache}>
                <InternalComponent />
            </CacheProvider>
        ));

        expect(providedCache).toEqual(cache);
    });
});