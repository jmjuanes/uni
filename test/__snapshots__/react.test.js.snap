// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[react] Box should render 1`] = `
<div
  className="uni-655233114"
>
  Hello world
</div>
`;

exports[`[react] CacheProvider should render 1`] = `
<div>
  Hello world
</div>
`;

exports[`[react] ThemeProvider should render 1`] = `
<div>
  Hello world
</div>
`;

exports[`[react] styled should return a valid React component 1`] = `
<div
  className="uni-655233114"
>
  Hello world
</div>
`;
